pkg/util/ratelimit/bucket.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratelimit
+
+import (
+	"sync"
+	"time"
+)
+
+// Bucket models a token bucket
+type Bucket struct {
+	unitsPerNano float64
+	nanosPerUnit float64
```

Comment:
Will do - I also hand-waving satisifed myself that there's no loss of precision for qps from about 1E6 to 1E-6.

I'll actually add some unit tests just to prove it :-)